### ISSUE TITLE
Unable to update handler.context via spread operator

### ISSUE THREAD

**Issue by OAFCROB** [middyjs/middy#484]
If you create an async middleware that needs to update the handler context for use within the handler, you're unable to update the context with a spread operator.

Please note this is done in a typescript context. 

```
import middy from "@middy/core";
import { Context } from "aws-lambda";
interface MiddlewareParameters {
 // Custom parameters
}
export interface CustomContext extends Context {
  dependencies: {
   test: string
  };
}

const testMiddleware: middy.Middleware<
  MiddlewareParameters,
  CustomContext
> = () => {
  return {
    before: async handler => {
      (handler.context as CustomContext) = {
        ...handler.context,
        dependencies: {
          test: "Set test dependency"
        }
      };
    }
  };
};

export default testMiddleware;
```

In my handler I was expecting that the Lambda's context would have been updated and show the dependencies but this wasn't the case. I was able to get the context to update as expected by altering the [context](https://github.com/middyjs/middy/blob/master/src/middy.js#L180) in middy to `instance.middy`. Is this as intended or is this a slight bug? 

I managed to eventually work around the issue with object.assign 
```
Object.assign(handler.context, {
        dependencies: {
          test: repository,
          test: "Set test dependency"
        }
      });
```

**Comment by Dartv** [comment]
Same issue here. Looks like we can't override context with typescript

**Comment by OAFCROB** [comment]
@Dartv we eventually managed to get this working for our use case, it might be of some use.


Custom middleware
```
import middy from "@middy/core";
import { Context, Handler, APIGatewayProxyEvent } from "aws-lambda";
import { withContext } from "../utilities/middyTypes";

export interface CustomContext extends Context {
  dependencies: {
    test: string;
  };
}

const customMiddleware = (): // We don't use the middle argument but have to use any instead of unknown because middy typing isn't right...
middy.MiddlewareObject<
  APIGatewayProxyEvent,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  any,
  CustomContext
> => {
  return {
    before: async (handler) => {
      handler.context.dependencies = {
        test: "hello world"
      };
    }
  };
};

// Helper to add custom context to the handler
export type withCustomContext<H extends Handler> = withContext<
  H,
  CustomContext
>;

export default customMiddleware;
```

MiddyTypes
```
import { Callback, Context, Handler } from "aws-lambda";
import middyLib from "@middy/core";

// Extends the Context type second argument
export type withContext<H extends Handler, C extends Context> = ExtendContext<
  Parameters<H>,
  ReturnType<H>,
  C
>;
// Ensures that there are 3 arguments and the middle one is the context to update
type ExtendContext<
  Args extends [unknown, Context, unknown],
  ReturnType,
  C extends Context
> = (event: Args[0], context: C, callback: Args[2]) => ReturnType;

// the middy lib isn't typed correctly so these are some type fixes
export type MiddyHandler<E, R, C> = (
  event: E,
  context: C,
  callback: Callback<R>
) => void | Promise<R>;
export const middy = middyLib as <E, R, C extends Context>(
  handler: MiddyHandler<E, R, C>
) => middyLib.Middy<E, R, C>;
```

**Comment by willfarrell** [comment]
I personally don't use ts. Is there something we need to add to the library or the documentation to resolve this?

**Comment by russell-dot-js** [comment]
Fix here: #556
